consensus/enclave/impl/src/lib.rs

```diff
@@ -697,20 +708,81 @@ impl ConsensusEnclave for SgxConsensusEnclave {
             &fee_public_key.view_public_key,
         );
 
-        let fee_outputs = total_fees_by_token_id
+        // Compute the total fees for each known token id, for tx's in this block.
+        let mut total_fees: CtTokenMap<u128> = ct_min_fee_map.keys().cloned().collect();
+        for tx in transactions.iter() {
+            let token_id = TokenId::from(tx.prefix.token_id);
+            total_fees.add(&token_id, tx.prefix.fee as u128);
+        }
+
+        // Sort the token ids which did not appear to the end.
+        total_fees.sort_zeroes_to_end();
+
+        // Make fee outputs.
+        //
+        // We always make a fee output for any fee amount which is nonzero.
+        // However, we sometimes make zero value fee outputs, for privacy.
+        // The number of fee outputs we make is min(num_tokens, num_transactions).
+        //
+        // That is, if there is only transaction output in the block, then we only make
+        // one fee output. If there are two transactions in the block, we have
+        // to make two fee ouptuts, even if they used the same token, to avoid
+        // revealing to these parties that the other person used the same token
+        // id as them (active attack). So in that case we will mint a zero value
+        // fee.
+        //
+        // However, we never have to make more fee outputs than there are tokens,
+        // no matter how many transactions there are in the block.
+        //
+        // This strategy is optimal both for blocks with one transaction, and blocks
+        // with many transactions -- it only creates some extra fee outputs in the
+        // mid-range, which seems okay.
+        //
+        // Note: The "take" here only makes sense because sort_zeros_to_end was called.
+        // This ensures that everything we are ignoring via take has zero value.
+        let fee_outputs = total_fees
             .iter()
-            .map(|(token_id, total_fee)| {
-                mint_output(
+            .take(min(total_fees.len(), transactions.len()))
+            .flat_map(|(token_id, total_fee)| {
+                // While unlikely, it is technically possible for the accumulated
+                // fee value to overflow a u64, particularly for custom tokens.
+                // Therefore, the accumulated fees are stored in a u128.
+                // Here we deal with the overflow
+                let mut outputs = Vec::default();
+
+                let token_id = *token_id;
+                let mut total_fee = *total_fee;
+
+                while total_fee > u64::MAX as u128 {
```

Comment:
You can avoid duplicating the `mint_output` call by changing the loop tobe
```
while total_fee > 0 {
    let output_fee = min(total_fee, u64::MAX);
    outputs.push(mint_output(...));
    total_fee -= output_fee;
}
```

Reply:
so i actually don't want to do it this way, because if the total fee is zero this doesn't mint anything, but it is desired to mint once even when total fee is zero

i'm using a `do...while` loop pattern here instead, because yeah its better to avoid writing `mint_output` twice